src/DecryptionFailureTracker.js

```diff
@@ -0,0 +1,124 @@
+/*
+Copyright 2018 New Vector Ltd
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+class DecryptionFailure {
+    constructor(failedEventId) {
+        this.failedEventId = failedEventId;
+        this.ts = Date.now();
+    }
+}
+
+export default class DecryptionFailureTracker {
+    // Array of items of type DecryptionFailure. Every `CHECK_INTERVAL_MS`, this list
+    // is checked for failures that happened > `GRACE_PERIOD_MS` ago. Those that did
+    // are added to `failuresToTrack`.
+    failures = [];
+
+    // Every TRACK_INTERVAL_MS (so as to spread the number of hits done on Analytics),
+    // one DecryptionFailure of this FIFO is removed and tracked.
+    failuresToTrack = [];
+
+    // Spread the load on `Analytics` by sending at most 1 event per
+    // `TRACK_INTERVAL_MS`.
+    static TRACK_INTERVAL_MS = 1000;
+
+    // Call `checkFailures` every `CHECK_INTERVAL_MS`.
+    static CHECK_INTERVAL_MS = 5000;
+
+    // Give events a chance to be decrypted by waiting `GRACE_PERIOD_MS` before moving
+    // the failure to `failuresToTrack`.
+    static GRACE_PERIOD_MS = 5000;
+
+    constructor(fn) {
+        if (!fn || typeof fn !== 'function') {
+            throw new Error('DecryptionFailureTracker requires tracking function');
+        }
+
+        this.trackDecryptionFailure = fn;
```

Comment:
This should probably have an underscore

Reply:
Also maybe a name with 'callback' in it or something, as having two functions called `trackDecryptionFailure` and `trackFailure` is a bit confusing.

Reply:
<del>In fact, the separation of concerns is nice, but in this case it seems, if anything, better to have `DecryptionFailureTracker` making pokes to analytics than `MatrixChat`, so I would probably just have this call `Analytics.trackEvent()` directly.</del>* Also since it has it's own start method, it seems like it may as well add its own `eventDecrypted` listener too? Hopefully this should still be fine for the unit tests since they'd continue to just not call `start()` but call the functions directly instead.

* Ah, again, because unit tests.

Reply:
>because unit tests

Yep. DFT doesn't depend on anything, which is a really nice property for testing.